products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,345 @@
+            self._database = database
+            self.delegate = delegate
+        }
+        
+        /// Returns the current iteration of the build
+        public func getCurrentIteration() throws -> UInt64 {
+            var success = false
+            
+            let errorPtr = MutableStringPointer()
+            let iteration = llb_database_get_current_iteration(_database, &success, &errorPtr.ptr)
+            if let error = errorPtr.msg {
+                throw Error.operationDidFail(error)
+            }
+            if !success {
+                throw Error.unknownError
+            }
+            return iteration
+        }
+        
+        /// Saves the current build iteration to the database
+        public func setCurrentIteration(_ iteration: UInt64) throws {
+            let errorPtr = MutableStringPointer()
+            llb_database_set_current_iteration(_database, iteration, &errorPtr.ptr)
+            if let error = errorPtr.msg {
+                throw Error.operationDidFail(error)
+            }
+        }
+        
+        /// Get the result for a given keyID
+        public func lookupRuleResult(keyID: KeyID) throws -> RuleResult? {
+            let errorPtr = MutableStringPointer()
+            var result = llb_database_result_t()
+            
+            let ruleKey = delegate.getKey(id: keyID)
+            
+            let stored = ruleKey.withCString { ruleKey in
+                return llb_database_lookup_rule_result(_database, keyID, ruleKey, &result, &errorPtr.ptr)
+            }
+            
+            if let error = errorPtr.msg {
+                throw Error.operationDidFail(error)
+            }
+            
+            if !stored {
+                return nil
+            }
+            
+            return RuleResult(result: result)
+        }
+        
+        /// Fetches all keys from the database, returns it and saves if as \see allKeys on the database object
+        public func getKeys() throws -> BuildDBFetchKeysResult {
+            let errorPtr = MutableStringPointer()
+            
+            let keys = UnsafeMutablePointer<OpaquePointer?>.allocate(capacity: 1)
+            
+            let success = llb_database_get_keys(_database, keys, &errorPtr.ptr)
+            
+            if let error = errorPtr.msg {
+                throw Error.operationDidFail(error)
+            }
+            if !success {
+                throw Error.unknownError
+            }
+            
+            guard let resultKeys = keys.pointee else {
+                throw Error.unknownError
+            }
+            
+            return BuildDBFetchKeysResult(result: resultKeys)
+        }
+        
+        /// Dumps an overview of the database to stdout
+        public func dump() {
+            llb_database_dump(_database)
+        }
+        
+    }
+    
+    public func startSession(_ operations: (BuildDB) throws -> Void) throws {
+        try buildStarted()
+        defer { buildComplete() }
```

Comment:
Eek. We really don't want to do this I think. Is there a reason this was required? Is it because the DB wouldn't allow certain accesses outside of a session? If so, we might want to consider how to handle read only queries differently.

Reply:
buildStarted() and buildEnded() starts/ends an exclusive transaction to the database, it might be worth renaming those in the C++ layer as well…

Reply:
Ya, agree.